Cache/Hashing.lean

```diff
@@ -60,14 +61,26 @@ def HashMemo.filterByFilePaths (hashMemo : HashMemo) (filePaths : List FilePath)
 /-- We cache the hash of each file and their dependencies for later lookup -/
 abbrev HashM := StateT HashMemo CacheM
 
-/-- Gets the file paths to Mathlib files imported on a Lean source -/
-def getFileImports (source : String) (pkgDirs : PackageDirs) : Array FilePath :=
-  let s := Lean.ParseImports.main source (Lean.ParseImports.whitespace source {})
-  let imps := s.imports.map (·.module.components |> .map toString)
-    |>.filter fun parts => match parts.head? with
-      | some head => pkgDirs.contains head
-      | none => false
-  imps.map (mkFilePath · |>.withExtension "lean")
+/--
+Read the imports from the raw file `content` and return an array of tuples
+`(module name, source file)`, one per import.
+
+Note: `fileName` is a string which is purely used for displaying an error message if
+parsing imports from `content` should fail. It is intended to be the file's name.
+-/
+def getFileImports (content : String) (fileName : String := "") :
+    CacheM <| Array (Name × FilePath) := do
+  let sp := (← read).srcSearchPath
+  let fileImports : Array Import ← Lean.parseImports' content fileName
```

Comment:
One note to explain things: the main line of code in `parseImports'` is exactly the first line deleted:

```lean
def parseImports' (input : String) (fileName : String) : IO (Array Lean.Import) := do
  let s := ParseImports.main input (ParseImports.whitespace input {}) -- this one!
  match s.error? with
  | none => return s.imports
  | some err => throw <| IO.userError s!"{fileName}: {err}"
```